masserstein/spectrum.py

```diff
@@ -534,22 +534,22 @@ def fuzzify_peaks(self, sd, step):
         Applies a gaussian filter to the peaks, effectively broadening them
         and simulating low resolution. Works in place, modifying self.
         The parameter step gives the distance between samples in m/z axis.
-        Note that after the filtering, the area below curve is equal to 1,
-        instead of the sum of 'peak' intensities!
+        After the filtering, the area below curve (not the sum of intensities!)
+        is equal to the sum of the input peak intensities.
         """
         new_mass = np.arange(self.confs[0][0] - 4*sd, self.confs[-1][0] + 4*sd, step)
-        new_intensity = np.zeros(len(new_mass))
-        lb = new_mass[0]
-        for x, y in self.confs:
-            xrnb = int((x-lb)//step)  # x's index in new_mass
-            xr = lb + step*xrnb
-            lnb = int((xr-x+4*sd)//step)   # nb of steps left of x to add gauss
-            xlb = xr - step*lnb
-            xrb = xr + step*lnb
-            xv = np.array([xlb + i*step for i in range(2*lnb + 2)])
-            nv = y*norm.pdf(xv, x, sd)
-            new_intensity[(xrnb-lnb):(xrnb+lnb+2)] += nv
-        self.confs = [(x, y) for x, y in zip(new_mass, new_intensity)]
+        A = new_mass[:,np.newaxis] - np.array([m for m,i in self.confs])
+        # we don't need to evaluate gaussians to far from their mean,
+        # from our perspective 4 standard deviations from the mean is the same
+        # as the infinity; this allows to avoid overflow as well:
+        A[np.abs(A) > 4*sd] = np.inf  
+        A **= 2
+        A /= (-2*sd**2)
+        A = np.exp(A)
+        new_intensity = A @ np.array([i for m,i in self.confs])  # matrix multiplication
+        new_intensity /= (np.sqrt(2*np.pi)*sd)
```

Comment:
I am not sure, but scipy.stats.norm.pdf may do work here? It's already imported.

Reply:
AFAIK this is a piece of speed-critical code, and the hand-written code here might be faster than generic code which must handle non-zero mean in the gaussian function (here we avoid one FP operation, a pointless addition of zero). As far as I'm concerned it's a non-blocker, unless someone bothers to performance-test both versions and prove that the scipy version is just as fast or faster (which might be the case: scipy might be smarter about vectorizing stuff than numpy).